pkg/reconciler/buildrun/buildrun.go

```diff
@@ -249,11 +251,12 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 
 			taskRunStatus := trCondition.Status
 
-			// check if we should delete the generated service account by checking the build run spec and that the task run is csomplete
+			// check if we should delete the generated service account by checking the build run spec and that the task run is complete
 			if resources.IsGeneratedServiceAccountUsed(buildRun) && (taskRunStatus == corev1.ConditionTrue || taskRunStatus == corev1.ConditionFalse) {
 				serviceAccount := &corev1.ServiceAccount{}
 				serviceAccount.Name = resources.GetGeneratedServiceAccountName(buildRun)
 				serviceAccount.Namespace = buildRun.Namespace
+				buildRun.Status.ServiceAccountName = &serviceAccount.Name
```

Comment:
At this point in time, the service account is already set in the status. It still makes sense to call the function to get the generated name so that in-flight BuildRuns that were created before this new feature is added, are handled correctly. Assuming that was your thinking, I suggest to slightly change the logic and make a comment for this

```suggestion
				serviceAccount.Name = buildRun.Status.ServiceAccountName
				if serviceAccount.Name == "" {
					// We have an old BuildRun that does not yet have Status.ServiceAccountName set
					serviceAccount.Name = resources.GetGeneratedServiceAccountName(buildRun)
					buildRun.Status.ServiceAccountName = &serviceAccount.Name
				}
				serviceAccount.Namespace = buildRun.Namespace
```

Reply:
@SaschaSchwarze0 make sense